File: java_implementation/SimpleRNG.java
public class SimpleRNG {

    public static class RNGState {
        private long data;
        public RNGState(){
            this(42);
        }
        public RNGState(long seed) {
            this.data = seed;
        }
    }

    private static long max_32bit = (long) Math.pow(2, 32);
    public static long rand_r(SimpleRNG.RNGState rng_state) {
        long num = rng_state.data;
        num ^= (num << 13) % max_32bit;
        num ^= (num >> 17) % max_32bit;
        num ^= (num << 5) % max_32bit;
        rng_state.data = num;
        return num;
    }

    private static RNGState global_rng_state = new RNGState(42);

    public static long rand() {
        return rand_r(global_rng_state);
    }

    public static void seed(long seed) {
        global_rng_state.data = seed;
    }

    public static void main(String[] args) {
        for(int x = 0; x < 10; ++x){
            System.out.println(Long.toString(rand()));
        }
    }
}

File: cpp_implementation/simplerng.cc
#include "simplerng.h"

uint32_t SimpleRNG::rand(uint32_t &rng_state)
{
    uint32_t num = rng_state;
    num ^= num << 13;
    num ^= num >> 17;
    num ^= num << 5;
    rng_state = num;

    return num;
}

SimpleRNG::SimpleRNG(uint32_t seed)
:
d_rng_state(seed)
{}

uint32_t SimpleRNG::rand()
{
    return rand(d_rng_state);
}

void SimpleRNG::seed(uint32_t seed)
{
    d_rng_state = seed;
}

